File: python/5-todo-list.py
# TODO List Application
# This script demonstrates the use of lists and dictionaries in Python

# Initialize an empty list to store todo items
todo_list = []

def add_todo(title, description, priority="medium"):
    """
    Add a new todo item to the list.
    
    Args:
        title (str): The title of the todo item
        description (str): A description of the todo item
        priority (str): The priority of the todo item (low, medium, high)
    """
    # Create a dictionary for the todo item
    todo_item = {
        "id": len(todo_list) + 1,
        "title": title,
        "description": description,
        "priority": priority,
        "completed": False
    }
    
    # Add the todo item to the list
    todo_list.append(todo_item)
    print(f"Added: {title}")

def list_todos():
    """Display all todo items in the list."""
    if not todo_list:
        print("No todos found. Your list is empty!")
        return
    
    print("\n=== YOUR TODO LIST ===")
    for item in todo_list:
        status = "✓" if item["completed"] else " "
        print(f"{status} [{item['id']}] {item['title']} - {item['description']} (Priority: {item['priority']})")

def mark_completed(todo_id):
    """
    Mark a todo item as completed.
    
    Args:
        todo_id (int): The ID of the todo item to mark as completed
    """
    for item in todo_list:
        if item["id"] == todo_id:
            item["completed"] = True
            print(f"Marked as completed: {item['title']}")
            return
    
    print(f"Todo with ID {todo_id} not found.")

def remove_todo(todo_id):
    """
    Remove a todo item from the list.
    
    Args:
        todo_id (int): The ID of the todo item to remove
    """
    for i, item in enumerate(todo_list):
        if item["id"] == todo_id:
            removed_item = todo_list.pop(i)
            print(f"Removed: {removed_item['title']}")
            return
    
    print(f"Todo with ID {todo_id} not found.")

def filter_by_priority(priority):
    """
    Filter todos by priority.
    
    Args:
        priority (str): The priority to filter by (low, medium, high)
    """
    filtered_todos = [item for item in todo_list if item["priority"] == priority]
    
    if not filtered_todos:
        print(f"No todos found with priority: {priority}")
        return
    
    print(f"\n=== TODOS WITH PRIORITY: {priority.upper()} ===")
    for item in filtered_todos:
        status = "✓" if item["completed"] else " "
        print(f"{status} [{item['id']}] {item['title']} - {item['description']}")

# Main program
print("=== TODO LIST APPLICATION ===")
print("This program helps you manage your todo list.")

while True:
    print("\nOptions:")
    print("1. Add a new todo")
    print("2. List all todos")
    print("3. Mark a todo as completed")
    print("4. Remove a todo")
    print("5. Filter todos by priority")
    print("6. Exit")
    
    choice = input("\nEnter your choice (1-6): ")
    
    if choice == "1":
        title = input("Enter todo title: ")
        description = input("Enter todo description: ")
        priority = input("Enter priority (low, medium, high) [default: medium]: ").lower()
        
        if priority not in ["low", "medium", "high"]:
            priority = "medium"
            
        add_todo(title, description, priority)
        
    elif choice == "2":
        list_todos()
        
    elif choice == "3":
        try:
            todo_id = int(input("Enter the ID of the todo to mark as completed: "))
            mark_completed(todo_id)
        except ValueError:
            print("Please enter a valid ID (number).")
            
    elif choice == "4":
        try:
            todo_id = int(input("Enter the ID of the todo to remove: "))
            remove_todo(todo_id)
        except ValueError:
            print("Please enter a valid ID (number).")
            
    elif choice == "5":
        priority = input("Enter priority to filter by (low, medium, high): ").lower()
        if priority in ["low", "medium", "high"]:
            filter_by_priority(priority)
        else:
            print("Invalid priority. Please enter low, medium, or high.")
            
    elif choice == "6":
        print("Goodbye!")
        break
        
    else:
        print("Invalid choice. Please enter a number between 1 and 6.") 

File: javascript/5-todo-list.js
/**
 * TODO List Application
 * 
 * A command-line todo list manager that demonstrates:
 * - Object-oriented programming concepts
 * - Array manipulation methods
 * - Async/await for user input
 * - Error handling
 * - Data filtering and sorting
 * 
 * Features:
 * - Add new todos with title, description, and priority
 * - List all todos with completion status
 * - Mark todos as completed
 * - Remove todos
 * - Filter todos by priority
 * - Persistent storage using arrays
 * 
 * @author Ibrahim
 * @version 1.0
 */

const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

// Initialize an empty array to store todo items
const todoList = [];

/**
 * Adds a new todo item to the list
 * @param {string} title - The title of the todo
 * @param {string} description - The description of the todo
 * @param {string} [priority="medium"] - The priority level (low, medium, high)
 */
function addTodo(title, description, priority = "medium") {
    const todoItem = {
        id: todoList.length + 1,
        title,
        description,
        priority,
        completed: false
    };
    
    todoList.push(todoItem);
    console.log(`Added: ${title}`);
}

/**
 * Displays all todos in the list
 */
function listTodos() {
    if (todoList.length === 0) {
        console.log("No todos found. Your list is empty!");
        return;
    }
    
    console.log("\n=== YOUR TODO LIST ===");
    todoList.forEach(item => {
        const status = item.completed ? "✓" : " ";
        console.log(`${status} [${item.id}] ${item.title} - ${item.description} (Priority: ${item.priority})`);
    });
}

/**
 * Marks a todo as completed
 * @param {number} todoId - The ID of the todo to mark as completed
 */
function markCompleted(todoId) {
    const todo = todoList.find(item => item.id === todoId);
    if (todo) {
        todo.completed = true;
        console.log(`Marked as completed: ${todo.title}`);
    } else {
        console.log(`Todo with ID ${todoId} not found.`);
    }
}

/**
 * Removes a todo from the list
 * @param {number} todoId - The ID of the todo to remove
 */
function removeTodo(todoId) {
    const index = todoList.findIndex(item => item.id === todoId);
    if (index !== -1) {
        const removedItem = todoList.splice(index, 1)[0];
        console.log(`Removed: ${removedItem.title}`);
    } else {
        console.log(`Todo with ID ${todoId} not found.`);
    }
}

/**
 * Filters and displays todos by priority
 * @param {string} priority - The priority level to filter by
 */
function filterByPriority(priority) {
    const filteredTodos = todoList.filter(item => item.priority === priority);
    
    if (filteredTodos.length === 0) {
        console.log(`No todos found with priority: ${priority}`);
        return;
    }
    
    console.log(`\n=== TODOS WITH PRIORITY: ${priority.toUpperCase()} ===`);
    filteredTodos.forEach(item => {
        const status = item.completed ? "✓" : " ";
        console.log(`${status} [${item.id}] ${item.title} - ${item.description}`);
    });
}

/**
 * Prompts the user with a question and returns their response
 * @param {string} question - The question to ask the user
 * @returns {Promise<string>} The user's response
 */
async function askQuestion(question) {
    return new Promise((resolve) => {
        readline.question(question, resolve);
    });
}

/**
 * Main application loop
 */
async function main() {
    console.log("=== TODO LIST APPLICATION ===");
    console.log("This program helps you manage your todo list.");
    
    while (true) {
        console.log("\nOptions:");
        console.log("1. Add a new todo");
        console.log("2. List all todos");
        console.log("3. Mark a todo as completed");
        console.log("4. Remove a todo");
        console.log("5. Filter todos by priority");
        console.log("6. Exit");
        
        const choice = await askQuestion("\nEnter your choice (1-6): ");
        
        if (choice === '6') {
            console.log("Goodbye!");
            readline.close();
            break;
        }
        
        switch (choice) {
            case '1':
                const title = await askQuestion("Enter todo title: ");
                const description = await askQuestion("Enter todo description: ");
                let priority = (await askQuestion("Enter priority (low, medium, high) [default: medium]: ")).toLowerCase();
                
                if (!['low', 'medium', 'high'].includes(priority)) {
                    priority = 'medium';
                }
                
                addTodo(title, description, priority);
                break;
                
            case '2':
                listTodos();
                break;
                
            case '3':
                const completeId = parseInt(await askQuestion("Enter the ID of the todo to mark as completed: "));
                if (!isNaN(completeId)) {
                    markCompleted(completeId);
                } else {
                    console.log("Please enter a valid ID (number).");
                }
                break;
                
            case '4':
                const removeId = parseInt(await askQuestion("Enter the ID of the todo to remove: "));
                if (!isNaN(removeId)) {
                    removeTodo(removeId);
                } else {
                    console.log("Please enter a valid ID (number).");
                }
                break;
                
            case '5':
                const filterPriority = (await askQuestion("Enter priority to filter by (low, medium, high): ")).toLowerCase();
                if (['low', 'medium', 'high'].includes(filterPriority)) {
                    filterByPriority(filterPriority);
                } else {
                    console.log("Invalid priority. Please enter low, medium, or high.");
                }
                break;
                
            default:
                console.log("Invalid choice. Please enter a number between 1 and 6.");
        }
    }
}

// Start the application
main().catch(console.error); 